speaker/layer2_controller.go

```diff
@@ -33,21 +34,25 @@ type layer2Controller struct {
 	mList     *memberlist.Memberlist
 }
 
+// Last known value of the active nodes known via memberlist, per Service IP.
+//
+// This is used to detect a cluster membership change since the last time we
+// looked at this Service. The outer key is the service IP since we need to
+// detect changes on a per IP basis.  The value is a map of the current
+// active Nodes in the same format as the return of getActiveNodes().
+var lastActiveNodes map[string]map[string]bool
```

Comment:
why keeping activenode per service? active node is cluster wide from memberlist, not per service. is it simpler to respam all services when active node changes?

Reply:
> The cluster membership tracking is done on a per-IP basis. We can not
do this globally, because the first time we see that cluster
membership has changed, we likely have not finished syncing all
services to figure out who is the new owner of each IP. By tracking
changes on a per-IP basis, we ensure we send the messages only after
confirming we are still the owner of that IP after the cluster
membership change.